src/data/polynomial/denominators.lean

```diff
@@ -0,0 +1,75 @@
+/-
+Copyright (c) 2020 Damiano Testa. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Damiano Testa
+-/
+import data.polynomial.erase_lead
+/-!
+# Denominators of evaluation of polynomials at ratios
+
+Let `i : R → K` be a homomorphism of semirings.  Assume that `K` is commutative.  If `a` and
+`b` are elements of `R` such that `i b ∈ K` is invertible, then for any polynomial
+`f ∈ polynomial R` the "mathematical" expression `b ^ f.nat_degree * f (a / b) ∈ K` is in
+the image of the homomorphism `i`.
+-/
+
+open polynomial finset
+
+section denoms_clearable
+
+variables {R K : Type*} [semiring R] [comm_semiring K] {i : R →+* K}
+variables {a b : R} {bi : K}
+-- TODO: use hypothesis (ub : is_unit (i b)) to work with localizations.
+
+/-- `denoms_clearable` formalizes the property that `b ^ N f (a / b)`
+does not have denominators, if the inequality `f.nat_degree ≤ N` holds.  In the
+implementation, we also use provide an inverse in the existential.
+-/
+def denoms_clearable (a b : R) (N : ℕ) (f : polynomial R) (i : R →+* K) :=
+  ∃ D : R, ∃ bi : K, bi * i b = 1 ∧ i D = i b ^ N * eval (i a * bi) (polynomial.map i f)
+
+lemma denoms_clearable_zero (N : ℕ) (a b : R) (bu : bi * i b = 1) :
+  denoms_clearable a b N 0 i :=
+⟨0, bi, bu, by simp only [eval_zero, ring_hom.map_zero, mul_zero, map_zero]⟩
+
+lemma denoms_clearable_C_mul_X_pow (N : ℕ) (a b : R) (bu : bi * i b = 1) :
+  ∀ (n : ℕ) (r : R), n ≤ N → denoms_clearable a b N (C r * X ^ n) i :=
+begin
+  refine λ n r nN, ⟨r * a ^ n * b ^ (N - n), bi, bu, _⟩,
```

Comment:
If you start your proof with a lambda, you might as well put more assumptions before the `:` in the statement.

Reply:
Thanks: I should have looked out for that!